update contributing guide (#198)
flatironinstitute/nemos: docs/developers_notes/03-base_regressor.md

# The Abstract Class `BaseRegressor`

`BaseRegressor` is an abstract class that inherits from `Base`, stipulating the implementation of number of abstract methods such as `fit`, `predict`, `score`. This ensures seamless assimilation with `scikit-learn` pipelines and cross-validation procedures.

!!! Example
    The current package version includes a concrete class named `nemos.glm.GLM`. This class inherits from `BaseRegressor`, which in turn inherits `Base`, since it falls under the " GLM regression" category. 

>     As any `BaseRegressor`, it **must** implement the `fit`, `score`, `predict` methods.

Reviewer: I don't understand this sentence. I think what you mean is either "As a `BaseRegressor`, it must implement..." or "Like any `BaseRegressor`, it implements..."

### Abstract Methods

> For subclasses derived from `BaseRegressor` to function correctly, they must implement the following:

Reviewer: this directly contradicts the above example, which only says that GLM must implement ft, score, and predict

1. `fit`: Adapt the model using input data `X` and corresponding observations `y`.
2. `predict`: Provide predictions based on the trained model and input data `X`.
3. `score`: Score the accuracy of model predictions using input data `X` against the actual observations `y`.
4. `simulate`: Simulate data based on the trained regression model.
5. `update`: Run a single optimization step, and stores the updated parameter and solver state. Used by stochastic optimization schemes.
6. `_predict_and_compute_loss`: Compute prediction and evaluates the loss function prvided the parameters and `X` and `y`.
7. `_check_params`: Check the parameter structure.
8. `_check_input_dimensionality`: Check the input dimensionality matches model expectation.
9. `_check_input_and_params_consistency`: Checks that the input and the parameters are consistent.

> 10. `_get_coef_and_intercept` and `_set_coef_and_intercept`: set and get model coefficient and intercept term.

Reviewer: worth pointing out if these are used by other, non-abstract methods of BaseRegressor (which you *don't* need to implement), probably specifying them. E.g., `_predict_and_compute_loss` is used during `instantiate_solver`

### Attributes

> Public attributes are stored as properties:

Reviewer: worth saying: when writing a new baseregressor, do you need to do anything with these or do they come along for free?

Reviewer 2: I added a comment

- `regularizer`: An instance of the `nemos.regularizer.Regularizer` class. The setter for this property accepts either the instance directly or a string that is used to instantiate the appropriate regularizer.
- `regularizer_strength`: A float quantifying the amount of regularization.
- `solver_name`: One of the `jaxopt` solver supported solvers, currently "GradientDescent", "BFGS", "LBFGS", "ProximalGradient" and, "NonlinearCG".
- `solver_kwargs`: Extra keyword arguments to be passed at solver initialization.
- `solver_init_state`, `solver_update`, `solver_run`: Read-only property with a partially evaluated `solver.init_state`, `solver.update` and, `solver.run` methods. The partial evaluation guarantees for a consistent API for all solver.

> !!! note "Sovlers"

Reviewer: ```suggestion
!!! note "Solvers"
```

    Solvers are typically optimizers from the `jaxopt` package, but in principle they could be custom optimization routines as long as they respect the `jaxopt` api (i.e., have a `run`, `init_state`, and `update` method with the appropriate input/output types).

>     We choose to rely on `jaxopt` because it provides a comprehensive set of robust, GPU accelerated, batchable and differentiable optimizers in JAX, that are highly customizable. In the future we will provide a number of custom solvers optimized for convex stochastic optimization.

Reviewer: ```suggestion
    We rely on `jaxopt` because it provides a comprehensive set of robust, GPU accelerated, batchable and differentiable optimizers in JAX, that are highly customizable. In the future we may provide a number of custom solvers optimized for convex stochastic optimization.
```
under-promise and over-deliver

## Contributor Guidelines

### Implementing Model Subclasses

When devising a new model subclass based on the `BaseRegressor` abstract class, adhere to the subsequent guidelines:

- **Must** inherit the `BaseRegressor` abstract superclass.
- **Must** realize the abstract methods, see above.
- **Should not** overwrite the `get_params` and `set_params` methods, inherited from `Base`.
- **May** introduce auxiliary methods for added utility.

> ## Glossary

Reviewer: I like this glossary -- should we have a more general glossary page that other places link to? solver and runner are probably only necessary for developers, but it's necessary for users to understand regularization and optimization  as well

Reviewer 2: I agree, it would fit in a doc page dedicated to regularizers. When we need to introduce the fact that different regularizers requires different solvers etc
